tools/node-cache/README.md

```diff
@@ -0,0 +1,92 @@
+## Overview
+
+This example allows for use of the [node-local-dns
+cache](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/dns/nodelocaldns),
+without requiring the kubelet `--cluster-dns` flag to be altered.  In
+particular, this means it can be used on GKE clusters.
+
+This is experimental, but feedback is welcome - please tag @justinsb.
+
+## Warning: Network policy
+
+If running with network policy, please see the [README for
+node-local-dns](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/dns/nodelocaldns#network-policy-and-dns-connectivity);
+network policy will likely need to be configured for the node-local-dns agent.
+
+## Other caveats
+
+* If the node-local-dns agent crashes, DNS resolution will not function on that
+  node until it is restarted.
+
+* This configuration must be removed before installing the node-local-dns cache.
+
+## How it works
+
+Normally the node-local-dns agent:
+* uses an unused IP (typically `169.254.20.10`)
+* the kubelet `--cluster-dns` flag is used to specify that pods should use that
+  IP address (`169.254.20.10`) as their DNS server
+* CoreDNS runs as a daemonset on every node, configured to listen to the
+  internal IP (`169.254.20.10`)
+* CoreDNS caches results locally, but forwards uncached queries upstream to the
+  "real" kube-dns service
+* The node-local-dns agent configures IP tables rules to avoid conntrack / NAT
+
+In this mode, we instead intercept the existing kube-dns service IP a few
+things:
+* We configure the node-local-dns agent to intercept the kube-dns service IP
+* kubelet is already configured to send queries to that service, by default
+* We configure a second service to act as the "real" kube-dns service
+* When the node-local-dns agent configures the kube-dns service IP to avoid
+  conntrack/NAT, this takes precedence over the normal DNS service routing.
+
+## Installation
+
+A script is provided, simply run `./install.sh`
+
+## Removal
+
+Removal is more complicated that installation.  We can remove the daemonset, and as
+part of pod shutdown the node-local-dns cache should remove the IP interception
+rules.  However, if something goes wrong with the removal, the IP interception rules
+will remain in place, but the node-local-dns cache will not be running to serve
+the intercepted traffic, and DNS lookup will be broken on that node.  However,
+restarting the machine will remove the IP interception rules, so if this is done
+as part of a cluster update the system will self-heal.
```

Comment:
Is this the only reason why we would not make this the normal way of using dns local cache and instead require explicitly changing kubelet --cluster-dns option ?

Reply:
I feel like trying this out, quickly since we are running into this a lot. What other drawbacks we can think of of this approach ?

Reply:
The main reason to modify --cluster-dns and use a new ip was because this approach of re-using service Ip does not work very well in ipvs mode of kube-proxy. In ipvs mode, kube-proxy creates a local interface on the host binding the service ip there. So packets destined to service ip will hit ipvs rules and get forwarded to pods. If ipvs mode does not apply in the setup, this approach will work well.

Reply:
i think we should include this information about its limitation in ipvs mode in a limitations/caveats section to inform users using that mode.